### PR TITLE
feat: crystallize docs

### PR DESCRIPTION
**Pull request by GeopJr**
This PR attempts to "crystallize" the extracted docs so they follow [the "spec"](https://crystal-lang.org/reference/1.3/syntax_and_semantics/documenting_code.html#markup).

So far:

- It adds a warning above codeblocks that contain code in a language (code blocks with no language set don't have it) (*Warning message copied from gtk-rs*)

Screenshot of docs showing WARNING ⚠️ The following code is in c ⚠️

- Tries to link methods, classes etc (eg. `[method@Gtk.TextBuffer.set_enable_undo]` => `Gtk::TextBuffer#enable_undo=` (link))

Screenshot of docs showing Gtk::TextBuffer#enable_undo= as link

Draft for now as it probably needs further testing, optimization, specs & catching edge cases.

### CODE REVIEW

**Comment by hugopl** [comment]
Nice, I'll try here today or this weekend. Maybe the logic to translate C function names to Crystallized names can be shared with the generator... but I need to check if with the current code this is easy to do this since the generator code still need more love and organization.

**Comment by hugopl** [comment]
I think it's good to be merged... the code that does the method name transformation could be shared with the generator, but this would require a refactor, so maybe later.

The CI fail is due to ameba that is failing on nightly build, it's ok.

**Comment by GeopJr** [comment]
Last second fix: `[property@Gtk.Orientable:orientation]` => `Gtk::Orientable#orientation`

**Comment by hugopl** [comment]
BTW I think this also deserve a test, you can add a comment in C code of any function in `test_subject.h`, or even create a function that does nothing like `Test::Subject#doc_test` and add this documentation things you want to test... then the crystal test file would just read the generated source code and grep for the expected results.

**Comment by GeopJr** [comment]
Done, let me know if I should change anything!

screenshot of generated and converted docs of the test typelib

**Comment by hugopl** [comment]
I spent some time today to put some :heart: into doc generator, maybe it can have a small conflict with this patch... but now it render docs for methods and signals, and the code is less ugly  :sweat_smile: 

https://github.com/hugopl/gi-crystal/commit/0cbb5f7e48f3cfe76944045c9a418a8b2b366e3e

**Comment by hugopl** [comment]
One thing that could help you on modifications in the generator is the `make compare` and `make recompare`.

The first one generate the code using the last stable version, then with the current version then compare the files.

The last one compare the current one with whatever is in `/tmp/gi-crystal/old`

Shards that depend on gi-crystal can just call `./bin/compare-api`

**Comment by GeopJr** [comment]
Looks awesome! Also fixed the conflict 👍

File: spec/docs_spec.cr [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+      "# getter: `Test::Subject#crystallize_docs`",
+      "# setter: `Test::Subject#crystallize_docs=`",
+      "# is: `Test::Subject#is_crystallize_docs?`",
+      "# initializer: `Test::Subject.new`",
+      "# WARNING: **⚠️ The following code is in c ⚠️**",
+    ]
+    test_subject = File.read("./src/auto/test-1.0/subject.cr")
+
+    expected_results.each do |doc|
+      test_subject.includes?(doc).should eq(true)
```
Comment (hugopl) [inline spec/docs_spec.cr:15]:
```suggestion
      test_subject.should contain(doc)
```

File: spec/libtest/test_subject.h [hunk 2]
```diff
@@ -51,6 +51,26 @@ typedef struct _TestStruct {
...
  *
  * Main class used to test all sort of things directly related or not to GObject.
+ * 
+ * Example docs for testing:
+ *
+ * getter: [method@Test.Subject.get_crystallize_docs]
+ * 
+ * setter: [method@Test.Subject.set_crystallize_docs]
+ * 
+ * is: [method@Test.Subject.is_crystallize_docs]
```
Comment (hugopl) [inline spec/libtest/test_subject.h:61]: git show show some extra white spaces here
image

I thought github would show then in the web interface... :confused:

Comment (hugopl) [reply spec/libtest/test_subject.h:61]: You can use `test_subject_is_bool` on this instead of create an empty funcion.

File: spec/libtest/test_subject.h [hunk 3]
```diff
@@ -51,6 +51,26 @@ typedef struct _TestStruct {
  * TestSubject:
  *
  * Main class used to test all sort of things directly related or not to GObject.
+ * 
+ * Example docs for testing:
+ *
+ * getter: [method@Test.Subject.get_crystallize_docs]
```
Comment (hugopl) [inline spec/libtest/test_subject.h:57]: You can use `test_subject_get_out_param` to test this, to avoid create an empty function.

File: spec/libtest/test_subject.h [hunk 4]
```diff
@@ -51,6 +51,26 @@ typedef struct _TestStruct {
...
+ * 
+ * setter: [method@Test.Subject.set_crystallize_docs]
```
Comment (hugopl) [inline spec/libtest/test_subject.h:59]: You can use `test_subject_set_str_list` to test this, to avoid create an empty function.

File: spec/libtest/test_subject.h [hunk 5]
```diff
@@ -373,4 +393,25 @@ void test_subject_raise_file_error2(TestSubject* self, int foo, GError** error);
...
+ */
+void test_subject_set_crystallize_docs(TestSubject *self, const gchar* data);
+
+/**
+ * test_subject_is_crystallize_docs: (get-property Boolean)
+ *
+ * Used to test documentation conversion to Crystal.
+ */
+void test_subject_is_crystallize_docs(TestSubject *self);
+
```
Comment (hugopl) [inline spec/libtest/test_subject.h:416]: Better to remove these and use the already existing functions mentioned in comments above, otherwise they would need also an empty implementation on test_subject.c to avoid missing symbols in the shared library.

**Review by hugopl — CHANGES_REQUESTED** [review]
Looks great! I just did minor comments, after this I think we can merge... and maybe release a new version with all these doc changes.
